### PR TITLE
Add dynamic features

### PR DESCRIPTION
**Pull request by alexandru-calinoiu**
This is a proposal to split the app into feature modules, the main benefit in this case it will make the development easier as individual contributors can focus their effort on a individual feature and own it.

A brief explainer of the features can be found [here](https://developer.android.com/guide/playcore/feature-delivery#customize_delivery)

A brief explanation of the modules;

- `app` is the com.android.application, which is needed to create the app bundle. It will also init the dependency injection and other libraries that require init at the app level
- `core` it will be used for serving network requests or accessing the database
- `commons` are  com.android.library and they hold common resources shared between features. Reusing layouts, resources views without needing to duplicate code
- `features` modules are com.android.dynamic-feature it ill hold code required for a feature all assets and other dependecies

### What does it fix?

Closes #XXX

### CODE REVIEW

File: commons/ui/src/main/kotlin/com/teacherworkout/commons/ui/extensions/NavigationExtension.kt [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+ * Issue tracking: https://issuetracker.google.com/issues/127932815
+ * Issue github: https://github.com/android/architecture-components-samples/issues/530
+ */
+@Suppress("UnsafeCast", "UnsafeCallOnNullableType", "LongMethod", "ComplexMethod")
+fun BottomNavigationView.setupWithNavController(
+    navGraphIds: List<Int>,
+    fragmentManager: FragmentManager,
+    containerId: Int,
+    intent: Intent
+): LiveData<NavController> {
```
Comment (alexandru-calinoiu) [inline commons/ui/src/main/kotlin/com/teacherworkout/commons/ui/extensions/NavigationExtension.kt:25]: Not happy with this workaround at all, I need to spend some time going through the reasoning in both of the issue and pr and figure out what to do.

Any feedback will be greatly appreciated.

Comment (lukstbit) [reply commons/ui/src/main/kotlin/com/teacherworkout/commons/ui/extensions/NavigationExtension.kt:25]: Isn't this issue with multiple back stacks [solved](https://medium.com/androiddevelopers/multiple-back-stacks-b714d974f134) in the newest alphas for navigation? This is why I used the latest alpha for the navigation dependency in my PR.

**Comment by lukstbit** [comment]
@alexandru-calinoiu I have no experience with dynamic feature modules so I can't really comment on the code.

My understanding is that feature modules are related to on demand download  from Google Play distribution which doesn't apply to our app(because it is small and we would get basically no benefit vs complexity). However, the navigation library doesn't seem to support normal modules and we kind of need to use feature modules if we use the navigation library.

File: .idea/compiler.xml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <project version="4">
   <component name="CompilerConfiguration">
-    <bytecodeTargetLevel target="11" />
+    <bytecodeTargetLevel target="13" />
```
Comment (lukstbit) [inline .idea/compiler.xml:4]: Isn't JDK 11 used in Android Studio?

File: buildSrc/build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    mavenCentral()
+    maven("https://oss.sonatype.org/content/repositories/snapshots")
+    maven("https://plugins.gradle.org/m2/")
+}
+
+kotlinDslPluginOptions {
+    experimentalWarning.set(false)
+}
+
+object PluginsVersions {
```
Comment (lukstbit) [inline buildSrc/build.gradle.kts:18]: We should probably extract this into its own file.

File: build.gradle.kts [hunk 4]
```diff
@@ -1,5 +1,6 @@
 // Top-level build file where you can add configuration options common to all sub-projects/modules.
 buildscript {
+    val kotlin_version by extra("1.5.10")
```
Comment (lukstbit) [inline build.gradle.kts:3]: We should use the same version as the plugin 1.5.20.

**Comment by AlexandraDamaschin** [comment]
Merging this for now and will fix the improvements in another PR.

**Review by AlexandraDamaschin — APPROVED** [review]
Does this improve software quality? Yes

**Pull request merged**